src/eloelo/elodisco/hero_pool_generators.rs

```diff
@@ -0,0 +1,668 @@
+        pairing_order.into_iter().chain(t_min..t_max).collect()
+    }
+}
+
+impl HeroPoolGenerator for TaggedHeroPool {
+    fn assign_heroes(
+        &mut self,
+        hero_pools: Vec<(PlayerInfo, Vec<Hero>)>,
+    ) -> HashMap<PlayerInfo, Vec<Hero>> {
+        let hero_pools_sets = hero_pools
+            .into_iter()
+            .map(|(name, heroes)| (name, HashSet::from_iter(heroes.iter().cloned())))
+            .collect_vec();
+        let radiant = hero_pools_sets
+            .iter()
+            .filter(|(p, _)| p.is_radiant)
+            .sorted_by_key(|(p, _)| -p.elo)
+            .collect_vec();
+        let dire = hero_pools_sets
+            .iter()
+            .filter(|(p, _)| !p.is_radiant)
+            .sorted_by_key(|(p, _)| -p.elo)
+            .collect_vec();
+        let max_hero_shown = hero_pools_sets
+            .iter()
+            .map(|(p, _)| p.number_of_heroes_shown)
+            .max()
+            .unwrap_or(3);
+        let pairing_order = Self::players_pairing(radiant.len(), dire.len());
+        let pairs_tag = (0..pairing_order.len())
+            .map(HeroTag::next_tag)
+            .collect_vec(); // rotate Core -> Support -> Carry -> ... for balanced team composition
+        for _ in 0..max_hero_shown {
+            for (pair_id, pair_tag) in zip(pairing_order.iter(), pairs_tag.iter()) {
+                for team in [&radiant, &dire] {
+                    if team.len() <= *pair_id {
+                        // not even number of players per team
+                        continue;
+                    }
+                    let (player, hero_pool) = team.get(*pair_id).unwrap();
+                    if self
+                        .hero_assignement
+                        .get(player)
+                        .and_then(|v| Some(v.len() == (player.number_of_heroes_shown as usize)))
+                        .is_some_and(identity)
+                    {
+                        continue;
+                    }
+                    if let Some(pick) = self.assign_tagged_hero(hero_pool, pair_tag) {
+                        self.hero_assignement
+                            .entry(player.clone())
+                            .or_default()
+                            .push(pick.clone());
+                        self.taken.insert(pick);
+                    }
+                }
+            }
+        }
+        self.hero_assignement.clone()
+    }
+
+    fn reroll_user_heroes(
+        &mut self,
+        username: &DiscordUsername,
+        hero_pool: Vec<Hero>,
+    ) -> Vec<Hero> {
+        if self.hero_assignement.is_empty() {
+            return Vec::new();
+        }
+        let player = if let Some(p) = self.hero_assignement.keys().find(|p| p.name == *username) {
+            p.clone()
+        } else {
+            return Vec::new();
+        };
+        let player_tag =
+            self.deduce_tag(&self.hero_assignement.get(&player).unwrap().get(0).unwrap());
```

Comment:
Use `[]` for vector.

I understood the previous function that picks for a given player may be rotated and contain various tags. Is reroll deliberately ignoring that or am I missing something?